src/System.Net.Http/src/System/Net/Http/HPack/HPackDecoder.cs

```diff
@@ -0,0 +1,398 @@
+                        else if (_integerDecoder.BeginDecode((byte)val, LiteralHeaderFieldWithIncrementalIndexingPrefix))
+                        {
+                            OnIndexedHeaderName(_integerDecoder.Value);
+                        }
+                        else
+                        {
+                            _state = State.HeaderNameIndex;
+                        }
+                    }
+                    else if ((b & LiteralHeaderFieldWithoutIndexingMask) == LiteralHeaderFieldWithoutIndexingRepresentation)
+                    {
+                        _index = false;
+                        var val = b & ~LiteralHeaderFieldWithoutIndexingMask;
+
+                        if (val == 0)
+                        {
+                            _state = State.HeaderNameLength;
+                        }
+                        else if (_integerDecoder.BeginDecode((byte)val, LiteralHeaderFieldWithoutIndexingPrefix))
+                        {
+                            OnIndexedHeaderName(_integerDecoder.Value);
+                        }
+                        else
+                        {
+                            _state = State.HeaderNameIndex;
+                        }
+                    }
+                    else if ((b & LiteralHeaderFieldNeverIndexedMask) == LiteralHeaderFieldNeverIndexedRepresentation)
+                    {
+                        _index = false;
+                        var val = b & ~LiteralHeaderFieldNeverIndexedMask;
+
+                        if (val == 0)
+                        {
+                            _state = State.HeaderNameLength;
+                        }
+                        else if (_integerDecoder.BeginDecode((byte)val, LiteralHeaderFieldNeverIndexedPrefix))
+                        {
+                            OnIndexedHeaderName(_integerDecoder.Value);
+                        }
+                        else
+                        {
+                            _state = State.HeaderNameIndex;
+                        }
+                    }
+                    else if ((b & DynamicTableSizeUpdateMask) == DynamicTableSizeUpdateRepresentation)
+                    {
+                        if (_integerDecoder.BeginDecode((byte)(b & ~DynamicTableSizeUpdateMask), DynamicTableSizeUpdatePrefix))
+                        {
+                            // TODO: validate that it's less than what's defined via SETTINGS
+                            _dynamicTable.Resize(_integerDecoder.Value);
+                        }
+                        else
+                        {
+                            _state = State.DynamicTableSizeUpdate;
+                        }
+                    }
+                    else
+                    {
+                        // Can't happen
+                        throw new HPackDecodingException($"Byte value {b} does not encode a valid header field representation.");
+                    }
```

Comment:
Maybe restructure this if/else if chain as a switch?

Reply:
Is that better?

Reply:
> Is that better?

Personally, I'd find it easier to understand, but that's just personal preference.  Technically, depending on the values involved here, the compiler may be able to implement it as a jump table.

Reply:
It's a prefix match. The possible prefix values are:
1
01
001
0001
0000

So, we could implement it as a 16-way switch. Is that what you were thinking? I agree that's probably better for perf, assuming the compiler can generate a jump table.

Reply:
That said, I'm inclined to not change this for now and defer until we have some perf data.  There are probably bigger fish to fry here; e.g. we're storing the result on the IntegerDecoder, which is probably not optimal (we should just return it directly).

I'll add a TODO.